### PR TITLE
Fix config migration regressions - longer locks on conflicting resources, honor cache config in gateway mode

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Merge initConfig logic to ConfigSys

## Motivation and Context
simplify config migration

## How Has This Been Tested?
Manually

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#95a16918-632a-4134-bdfc-4d3a26e646b4) |
<a name="95a16918-632a-4134-bdfc-4d3a26e646b4"></a>
### 6312-52c5d3d/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31896
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp f34e3c61781d:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 3 minutes and 1 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 1 seconds
(4/13) Running aws-sdk-php tests ... done in 7 minutes and 31 seconds
(5/13) Running aws-sdk-ruby tests ... done in 39 seconds
(6/13) Running mc tests ... done in 5 minutes and 39 seconds
(7/13) Running minio-dotnet tests ... done in 6 minutes and 2 seconds
(8/13) Running minio-go tests ... FAILED in 8 minutes and 5 seconds
{
  "args": {
    "policy": "{\"expiration\":\"2018-08-29T03:12:39.302Z\",\"conditions\":[[\"eq\",\"$bucket\",\"minio-go-test-9hp4rotbzt6b0nr9\"],[\"eq\",\"$key\",\"o0mdcz6kdyzqgjcnzauzix3b39hewk\"],[\"eq\",\"$Content-Type\",\"binary/octet-stream\"],[\"eq\",\"$x-amz-meta-90par1c20npgltgctajo51fqm33q5x\",\"sdhhsd2qy4v2nr1yzjalqy6uqwjqzm\"],[\"content-length-range\", 10, 1048576]]}"
  },
  "duration": 427,
  "error": "400 Bad Request",
  "function": "PresignedPostPolicy(policy)",
  "message": "Http request failed",
  "name": "minio-go: testPresignedPostPolicy",
  "status": "FAIL"
}

Executed 7 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/config-current_test.go [hunk 1]
```diff
@@ -97,7 +97,7 @@ func TestServerConfigWithEnvs(t *testing.T) {
 	serverHandleEnvVars()
 
 	// Init config
-	initConfig()
+	initConfig(objLayer)
```
Comment (kannappanr) [inline cmd/config-current_test.go:100]: we should probably check for err here in this test.

File: cmd/config-migrate.go [hunk 2]
```diff
@@ -2412,7 +2412,7 @@ func migrateV27ToV28() error {
 // Migrates '.minio.sys/config.json' v27 to v28.
 func migrateMinioSysConfig(objAPI ObjectLayer) error {
 	// Construct path to config.json for the given bucket.
-	configFile := path.Join(bucketConfigPrefix, minioConfigFile)
```
Comment (kannappanr) [inline cmd/config-migrate.go]: was it not migrating properly before? As we are changing the bucket name here

File: cmd/config-current_test.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/config-current_test.go:100]: that is not what we are verifying here... so I didn't change the original intention.

File: cmd/config-migrate.go (hunk 2 above)
Comment (harshavardhana) [reply cmd/config-migrate.go]: we are I just changed it to the right value, it was a typo.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/config-current_test.go (hunk 1 above)
Comment (kannappanr) [reply cmd/config-current_test.go:100]: ok

File: cmd/config-current.go [hunk 3]
```diff
@@ -64,6 +64,12 @@ func (s *serverConfig) SetRegion(region string) {
 
 // GetRegion get current region.
 func (s *serverConfig) GetRegion() string {
+	if s == nil {
+		return ""
+	}
+	if globalIsEnvRegion {
```
Comment (kannappanr) [inline cmd/config-current.go]: should env variable take precedence even if `s` is `nil`

Comment (harshavardhana) [reply cmd/config-current.go]: yes you are right.

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#9d74e185-c8c0-4bfe-b08b-9878cc884f38) |
| mint-gateway-nas.sh | :x: [more...](#a5780911-8c58-43ed-a345-1605b3499422) |
<a name="9d74e185-c8c0-4bfe-b08b-9878cc884f38"></a>
### 6312-629ded4/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31367
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp a3f027016551:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 2 minutes and 38 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 7 minutes and 27 seconds
(5/13) Running aws-sdk-ruby tests ... done in 39 seconds
(6/13) Running mc tests ... done in 6 minutes and 1 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 51 seconds
(8/13) Running minio-go tests ... FAILED in 8 minutes and 25 seconds
{
  "args": {
    "policy": "{\"expiration\":\"2018-08-29T10:22:21.038Z\",\"conditions\":[[\"eq\",\"$bucket\",\"minio-go-test-05v1eni9tahog6y4\"],[\"eq\",\"$key\",\"kgstfufirsgvii5ecxm13bxfks96lf\"],[\"eq\",\"$Content-Type\",\"binary/octet-stream\"],[\"eq\",\"$x-amz-meta-15iohhrxwzyga2vi6wad9bz2ogwgn9\",\"tnhngpywmg9o3fe0lnp3bdmceeywui\"],[\"content-length-range\", 10, 1048576]]}"
  },
  "duration": 356,
  "error": "400 Bad Request",
  "function": "PresignedPostPolicy(policy)",
  "message": "Http request failed",
  "name": "minio-go: testPresignedPostPolicy",
  "status": "FAIL"
}

Executed 7 out of 13 tests successfully.
```
<a name="a5780911-8c58-43ed-a345-1605b3499422"></a>
### 6312-629ded4/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32010
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7db068891f9e:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 47 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 3 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... FAILED in 1 seconds
{
  "name": "mc",
  "duration": "342",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "/mint/run/core/mc/mc --config-folder /tmp/.mc-12872 --quiet --no-color cp /mint/data/datafile-65-MB myminio/mc-test-bucket-2959/mc-test-object-28592  >>> \n`/mint/data/datafile-65-MB` -> `myminio/mc-test-bucket-2959/mc-test-object-28592`\nmc: <ERROR> Failed to copy `/mint/data/datafile-65-MB`. Put http://minikube:32010/mc-test-bucket-2959/mc-test-object-28592?partNumber=1&uploadId=22fc4850-4d23-4c12-acc5-5bec786e33c5: net/http: HTTP/1.x transport connection broken: write tcp 172.17.0.26:47436->192.168.122.26:32010: write: connection reset by peer\nmc: <ERROR> Session safely terminated. To resume session `mc session resume SeXlYrTC`"
}

Executed 5 out of 13 tests successfully.
```

File: cmd/config.go [hunk 4]
```diff
@@ -229,39 +217,45 @@ func migrateConfigToMinioSys() error {
...
+		return errServerNotInitialized
+	}
+
 	if globalEtcdClient != nil {
 		ctx, cancel := context.WithTimeout(context.Background(), 20*time.Second)
 		resp, err := globalEtcdClient.Get(ctx, getConfigFile())
 		cancel()
 		if err == nil && resp.Count > 0 {
-			logger.FatalIf(migrateConfig(), "Config migration failed")
+			return migrateConfig()
```
Comment (nitisht) [inline cmd/config.go:230]: with this we'll be losing granular information on what failed during `initConfig`, why not keep `logger.FatalIf()`

Comment (harshavardhana) [reply cmd/config.go:230]: Structurally its not needed, the change is in line with all other functions for ConfigSys.

**Review by nitisht — APPROVED** [review]

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**